### ISSUE TITLE
IDE: Failed assert: "QSortFilterProxyModel: index from wrong model passed to mapFromSource"

### ISSUE THREAD

**Issue by brianlheim** [supercollider/supercollider#2934]
This one has been bugging me for awhile and since it seems there might possibly be related bugs popping up, I decided to give it its own issue.

# Problem

While debugging the SC IDE, the following error message appears in the console every time a new document is opened (including on startup):

```
QSortFilterProxyModel: index from wrong model passed to mapFromSource
```

# Diagnosis

Research: https://stackoverflow.com/questions/18364632/qsortfilterproxymodel-crashing-app

No clear answers yet, I think it's being caused by `mDocumentModel` in [this area](https://github.com/supercollider/supercollider/blob/master/editors/sc-ide/core/doc_manager.cpp#L272-L306) of code.

**Comment by brianlheim** [comment]
Forgot to say, I think this is potentially related to #2933, #2797, #2823. It looks like what's happening could be undefined behavior, which would explain the erratic nature of these reported bugs and why they are reproducible on some systems but not others.

**Comment by brianlheim** [comment]
Traced it to this line: https://github.com/supercollider/supercollider/blob/master/editors/sc-ide/widgets/editor_box.cpp#L151

**Comment by brianlheim** [comment]
So from git blame it looks like this was introduced by #2555 (e61dbfe).

**Comment by brianlheim** [comment]
This assert also fails when using drop-downs in the IDE.

**Comment by brianlheim** [comment]
@LucaDanieli @muellmusik @scztt can you take a look at this if you have time? You developed the original code, and I can only spend so much time reading up on Qt documentation :)

**Comment by muellmusik** [comment]
Err, I confess I haven't really dug through and re-familiarised myself with this code (sorry we're in the final push of exams period here!), so it's possible I've misunderstood something important, but...

Changing https://github.com/supercollider/supercollider/blob/master/editors/sc-ide/widgets/editor_box.cpp#L150 to this:

```QModelIndex mIndex = Main::documentManager()->docModel()->index(modelIndex, 0, QModelIndex());```

seems to solve the error anyway. This seemed an obvious candidate to me as ```Main::documentManager()->docModel()``` is the 'source' we are mapping 'from'. (in the constructor: ```mProxyModel->setSourceModel(Main::documentManager()->docModel());```)

Does that make sense? @LucaDanieli?I couldn't seem to reproduce the other issues you referenced, so I can't say if it's solved them.

If we're happy I can make a PR. (Or just commit to master?)

**Comment by LucaDanieli** [comment]
It seems not, as `index()` sets `QModelIndex()` as source by default:

`QModelIndex QSortFilterProxyModel::index(int row, int column, const QModelIndex &parent = QModelIndex()) const`

Maybe, could it be that `mProxyModel->setSourceModel(Main::documentManager()->docModel());` wants a `QAbstractItemModel` but `docModel()` returns a `QStandardItemModel`? So it warns that the model seems wrong as it is of the wrong type?

**Comment by muellmusik** [comment]
> It seems not, as index() sets QModelIndex() as source by default:
> 
> QModelIndex QSortFilterProxyModel::index(int row, int column, const QModelIndex &parent = QModelIndex()) const

I think you're reading that wrong? It creates an empty QModelIndex as parent by default. Also I don't think parent and source are equivalent here? (But I'm not that immersed in Qt-ness at the mo...)

**Comment by muellmusik** [comment]
Yes, a quick look at the doc suggests that the parent field here is optional as child items can have a parent index, but not all items need to.

**Comment by brianlheim** [comment]
Closed by #2942.
